System/Random.hs

```diff
@@ -20,41 +20,181 @@
---   data type 'StdGen'.  Programmers may, of course, supply their own
---   instances of 'RandomGen'.
+-- As an example, @rolls@ generates @n@ random values of @Word8@ in the range
+-- @[1, 6]@.
 --
--- * The class 'Random' provides a way to extract values of a particular
---   type from a random number generator.  For example, the 'Float'
---   instance of 'Random' allows one to generate random values of type
---   'Float'.
+-- >>> :{
+-- let rolls :: MonadRandom g m => Int -> g -> m [Word8]
+--     rolls n = replicateM n . uniformR (1, 7)
+-- :}
+--
+-- Given a /monadic/ pseudo-random number generator, you can run this
+-- probabilistic computation as follows:
+--
+-- >>> monadicGen <- MWC.create
+-- >>> rolls 10 monadicGen :: IO [Word8]
+-- [2,3,6,6,4,4,3,1,5,4]
+--
+-- Given a /pure/ pseudo-random number generator, you can run it in an 'IO' or
+-- 'ST' context using 'runPrimGenIO' or 'runPrimGenST' and their variants,
+-- respectively. If the pseudo-random number generator does not implement
+-- 'Prim', you can also use 'runMutGenIO' or 'runMutGenST' and their variants.
+--
+-- >>> let pureGen = mkStdGen 42
+-- >>> runPrimGenIO_ pureGen (rolls 10) :: IO [Word8]
+-- [1,1,3,2,4,5,3,4,6,2]
+--
+-- = How to generate random values in pure code
+--
+--  In pure code, use 'runGenState' and its variants to extract the pure random
+--  value from a monadic computation based on a pure pseudo-random number
+--  generator.
 --
--- This implementation uses the SplitMix algorithm [1].
+-- >>> let pureGen = mkStdGen 42
+-- >>> runGenState_ pureGen (rolls 10) :: [Word8]
+-- [1,1,3,2,4,5,3,4,6,2]
 --
--- [/Example for RNG Implementors:/]
+-- = How to implement 'RandomGen'
 --
--- Suppose you want to use a [permuted congruential
--- generator](https://en.wikipedia.org/wiki/Permuted_congruential_generator)
--- as the source of entropy (FIXME: is that the correct
--- terminology). You can make it an instance of `RandomGen`:
+-- Consider these points when writing a 'RandomGen' instance for a given pure
+-- pseudo-random number generator:
+--
+-- *   If the pseudo-random number generator has a power-of-2 modulus, that is,
+--     it natively outputs @2^n@ bits of randomness for some @n@, implement
+--     'genWord8', 'genWord16', 'genWord32' and 'genWord64'. See below for more
+--     details.
+--
+-- *   If the pseudo-random number generator does not have a power-of-2
+--     modulus, implement 'next' and 'genRange'. See below for more details.
+--
+-- *   If the pseudo-random number generator is splittable, implement 'split'.
+--
+-- Additionally, implement 'Prim' for the pseudo-random number generator if
+-- possible. This allows users to use the fast 'MutGen' adapter with the
+-- pseudo-random number generator.
+--
+-- == How to implement 'RandomGen' for a pseudo-random number generator with power-of-2 modulus
+--
+-- Suppose you want to implement a
+-- <https://en.wikipedia.org/wiki/Permuted_congruential_generator permuted
+-- congruential generator>.
```

Comment:
I don't know why but this link doesn't render properly in your picture.

Reply:
I find it easier to use markdown style links in haddock, so this will work
`[permuted congruential generator](https://en.wikipedia.org/wiki/Permuted_congruential_generator)`

Reply:
Ah, I didn't know one can use Markdown-style links in Haddock. Fixed.